packages/firestore/src/api/database.ts

```diff
@@ -2496,6 +2646,30 @@ function resultChangeType(type: ChangeType): firestore.DocumentChangeType {
   }
 }
 
+/**
+ * Converts custom model object of type T into DocumentData by applying the
+ * converter if it exists.
+ *
+ * This function is used when converting user objects to DocumentData
+ * because we want to provide the user with a more specific error message if
+ * their set() or fails due to invalid data originating from a toFirestore()
+ * call.
+ */
+function applyFirestoreDataConverter<T>(
+  converter: firestore.FirestoreDataConverter<T> | undefined,
+  value: T,
+  functionName: string
+): [firestore.DocumentData, string] {
+  let convertedValue;
+  if (converter) {
+    convertedValue = converter.toFirestore(value);
+    functionName = 'toFirestore() in Transaction.set';
```

Comment:
This is surprising to me (I wouldn't expect the error message to contain a Transaction string if I use a WriteBatch). Can this be done at the callsite?

Reply:
Oops, typo. Meant to add the functionName that is passed in, so in this case, `WriteBatch.set()` would be passed in if the user used a WriteBatch.